File: Hito_4/Observer/EjemploDetallado/VehicleRoutingOriginal/VehicleRoutingSolucionadoObserver/VehicleRoutingValidator/Handlers/FirstAndLastLocationsAreTheDepotHandler.cs
using VehicleRoutingValidator.VehicleRouting;

namespace VehicleRoutingValidator.Handlers;

public class FirstAndLastLocationsAreTheDepotHandler: BaseHandler
{
    private const string Description = "Solution does not start or end at depot.";
    private const int Depot = VehicleRoutingProblem.Depot;
    
    public override bool IsSolutionValid(int[] solution)
    {
        if(DoesSolutionStartAndEndAtDepot(solution))
            return base.IsSolutionValid(solution);
        UpdateSubscribers(Description);
        return false;
    }

    private bool DoesSolutionStartAndEndAtDepot(int[] solution)
        => DoesSolutionStartAtDepot(solution) && DoesSolutionEndAtDepot(solution);

    private bool DoesSolutionStartAtDepot(int[] solution)
        => solution[0] == Depot;

    private bool DoesSolutionEndAtDepot(int[] solution)
        => solution[^1] == Depot;

}

File: Hito_4/Observer/Tests_positivos/Test_Detallado_VehicleRouting/handlers/FirstAndLastLocationsAreTheDepotHandler.py
import BaseHandler
from ..models.VehicleRoutingProblem import VehicleRoutingProblem

class FirstAndLastLocationsAreTheDepotHandler(BaseHandler):
    DESCRIPTION = "Solution does not start or end at depot."
    DEPOT = VehicleRoutingProblem.DEPOT

    def __init__(self):
        super().__init__()

    def is_solution_valid(self, solution: list[int]) -> bool:
        if self._does_solution_start_and_end_at_depot(solution):
            return super().is_solution_valid(solution)
        self.update_subscribers(self.DESCRIPTION)
        return False

    def _does_solution_start_and_end_at_depot(self, solution: list[int]) -> bool:
        return self._does_solution_start_at_depot(solution) and self._does_solution_end_at_depot(solution)

    def _does_solution_start_at_depot(self, solution: list[int]) -> bool:
        return solution[0] == self.DEPOT

    def _does_solution_end_at_depot(self, solution: list[int]) -> bool:
        return solution[-1] == self.DEPOT
